src/embedding.jl

```diff
@@ -0,0 +1,172 @@
+"""
+    embedGigaSOM(som::Som, data::DataFrame, k, adjust, smooth)
+
+Return a data frame with X,Y coordinates of EmbedSOM projection of the data.
+
+# Arguments
+- `som`: a trained SOM
+- `data`: Array or DataFrame with the data.
+- `k`: number of nearest neighbors to consider (high values get quadratically slower)
+- `adjust`: position adjustment parameter (higher values avoid non-local approximations)
+- `smooth`: approximation smoothness (the higher the value, the larger the neighborhood of approximate local linearity of the projection)
+
+Example:
+
+Produce a 2-column matrix with 2D cell coordinates:
+
+```
+e = embedGigaSOM(som, data)
+```
+
+Plotting is best done using histograms; e.g. with Gadfly:
+
+```
+draw(PNG("output.png",20cm,20cm),
+     plot(x=e[:,1], y=e[:,2],
+     Geom.histogram2d(xbincount=200, ybincount=200)))
+```
+
+Data must have the same number of dimensions as the training dataset
+and will be normalised with the same parameters.
+"""
+function embedGigaSOM(som::GigaSOM.Som, data::DataFrame;
+                      knnTreeFun = BruteTree,
+                      k=0, adjust=1.0, smooth=0.0)
+
+    data::Array{Float64,2} = convertTrainingData(data)
+    if size(data,2) != size(som.codes,2)
+        println("    data: $(size(data,2)), codes: $(size(som.codes,2))")
+        error(SOM_ERRORS[:ERR_COL_NUM])
+    end
+
+    boost = ((1+sqrt(Float64(5)))/2)^(Float64(smooth)-2)
+    adjust = Float64(adjust)
+
+    if k == 0
+        k = Integer((som.xdim+som.ydim)/2)
+    end
+
+    if k > som.xdim*som.ydim
+        k = Integer(som.xdim * som.ydim)
+    end
+
+    t = knnTreeFun(Array{Float64,2}(transpose(som.codes)))
+
+    ndata = size(data, 1)
+    dim = size(data, 2)
+
+    nWorkers = nprocs()
+    if nWorkers > 1
+        dData = distribute(data)
+
+        dRes = [ (@spawnat w embedGigaSOM_internal(som, localpart(dData), t, k, adjust, boost)) for w in workers() ]
+
+        #hopefully the data are separated to localparts in correct order...
+        return vcat([fetch(r) for r in dRes]...)
```

Comment:
nice! 👍 The order will be ok if the workers have been enumerated (see comment above)

Reply:
Good to know, thanks a lot!